### DISCUSSION TITLE
permission denied and timeout when using systemctl service

### DISCUSSION THREAD

**Discussion by 07freedom** [transmission/transmission#5636] [Q&A]
### What is the issue?

Environment: Ubuntu 20
Version:  4.0.3

Description:
I did the installation from tar.gz file and when I try this command to boot, transmission works just fine.
```
transmission-daemon
```
The problem occures when I try to set a transmission-daemon.service to start it on boot.
The file is /lib/systemd/system/transmission-daemon.service
```
[Unit]
Description=Transmission BitTorrent Daemon
Wants=network-online.target
After=network-online.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/local/bin/transmission-daemon -f --log-level=error --config-dir /home/xxx/.config/transmission-daemon
ExecReload=/bin/kill -s HUP $MAINPID

[Install]
WantedBy=multi-user.target
```

Service fails to start and the log goes:
```
Jun 17 17:45:17 Gen10 transmission-daemon[11375]: [2023-06-17 17:45:17.723] ERR utils.cc:72 Couldn't read '/home/xxx/.config/transmission-daemon/settings.json': Permission denied (13) (utils.cc:72)
Jun 17 17:45:17 Gen10 transmission-daemon[11375]: [2023-06-17 17:45:17.723] ERR variant.cc:1098 Couldn't save '/home/xxx/.config/transmission-daemon/settings.json': Permission denied (13) (variant.cc:1098)
Jun 17 17:45:17 Gen10 transmission-daemon[11375]: [2023-06-17 17:45:17.723] ERR variant.cc:1098 Couldn't save '/home/xxx/.config/transmission-daemon/bandwidth-groups.json': Permission denied (13) (variant.cc:1098)
Jun 17 17:46:05 Gen10 nextcloud.mdns-publisher[1739]: 2023/06/17 17:46:05 WARNING: interface "enp2s0f0" has no IP addresses
Jun 17 17:46:46 Gen10 systemd[1]: transmission-daemon.service: start operation timed out. Terminating.
Jun 17 17:46:46 Gen10 transmission-daemon[11375]: [2023-06-17 17:46:46.890] ERR utils.cc:72 Couldn't read '/home/xxx/.config/transmission-daemon/settings.json': Permission denied (13) (utils.cc:72)
Jun 17 17:46:46 Gen10 transmission-daemon[11375]: [2023-06-17 17:46:46.890] ERR variant.cc:1098 Couldn't save '/home/xxx/.config/transmission-daemon/settings.json': Permission denied (13) (variant.cc:1098)
Jun 17 17:46:46 Gen10 transmission-daemon[11375]: [2023-06-17 17:46:46.890] ERR variant.cc:1098 Couldn't save '/home/xxx/.config/transmission-daemon/bandwidth-groups.json': Permission denied (13) (variant.cc:1098)
Jun 17 17:46:46 Gen10 transmission-daemon[11375]: Closing transmission session... done.
Jun 17 17:46:46 Gen10 systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
```
I have already tried

- change the folder owner to **debian-transmission** 
- change **privilege**  to 777
- put config folder to other location like /var/lib/transmission-daemon/.config/transmission-daemon/

but it didn't work. All I can do is to start it manully everytime I boot it . 
Any ideas how to fix that? Also do I have to run transmission with user debian-transmission?

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.3 Release

**Comment by Daniel6372** [comment]
Had the same issue running on arch, I don't think I was able to get it working either.
As a workaround you can make it run as your own user, presumably the one you ran the transmission-daemon command as, although this is considered less secure. Create the necessary directory and file:

`/etc/systemd/system/transmission-daemon.service.d/transmission.conf`
```
[Service]
User=youruser
```

**Comment by killemov** [comment]
My `Service` section:
```
[Service]
User=debian-transmission
TimeoutStartSec=infinity
Type=notify
ExecStart=/usr/bin/transmission-daemon -f --log-error
ExecStop=/bin/kill -s STOP $MAINPID
ExecReload=/bin/kill -s HUP $MAINPID
Restart=always
Environment=TRANSMISSION_WEB_HOME=/data/transfer/shift
NoNewPrivileges=true
```

Please note that every part of `/home/xxx/.config/transmission-daemon/` should be readable by transmission-daemon. Test this by becoming transmission-daemon and then reading the configuration files.

**Comment by holta** [comment]
Just for comparison Ubuntu 23.10 pre-releases use this Transmission 4.0.2 systemd unit file, if you run `apt install transmission-daemon` :

```
# /lib/systemd/system/transmission-daemon.service
[Unit]
Description=Transmission BitTorrent Daemon
Wants=network-online.target
After=network-online.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/bin/transmission-daemon -f --log-level=error
ExecReload=/bin/kill -s HUP $MAINPID
NoNewPrivileges=true
MemoryDenyWriteExecute=true
ProtectSystem=true
PrivateTmp=true

[Install]
WantedBy=multi-user.target
```

```
# apt list transmission*
Listing... Done
transmission-cli/mantic,now 4.0.2-1ubuntu2 amd64 [installed,automatic]
transmission-common/mantic,now 4.0.2-1ubuntu2 amd64 [installed,automatic]
transmission-daemon/mantic,now 4.0.2-1ubuntu2 amd64 [installed]
transmission-gtk/mantic 4.0.2-1ubuntu2 amd64
transmission-qt/mantic 4.0.2-1ubuntu2 amd64
transmission-remote-gtk/mantic 1.5.1-1 amd64
transmission/mantic 4.0.2-1ubuntu2 amd64
```

Tangentially:

- #5362
- #5585

**Comment by lindianfeng** [comment]
please install libsystemd-dev and rebuild

**Reply by 07freedom** [reply]
For the rebuild, do I have to delete the former bin like transmission-daemon in /usr/local/bin/ or just do "sudo make install"?

**Reply by GaryElshaw** [reply]
sudo wil be fine, just make sure you start with a clean build directory.

**Reply by 07freedom** [reply]
Thanks for the tips but it still fail to start with the same error

**Comment by ingria** [comment]
I have the same error but without "permission denied". 

```
[2023-08-01 18:28:46.487] inf session.cc:1328 Transmission version 4.0.3 (6b0e49bbb2) shutting down (session.cc:1328)
[2023-08-01 18:28:46.487] inf rpc-server.cc:740 Stopped listening for RPC and Web requests on '0.0.0.0:9091' (rpc-server.cc:740)
transmission-daemon[574931]: Closing transmission session... done.
systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
systemd[1]: Failed to start Transmission BitTorrent Daemon.
```

It works fine while systemd is trying to launch the daemon, but then it gets killed by systemd because of timeout.

I'm using the following service file:

```
[Unit]
Description=Transmission BitTorrent Daemon
After=network.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/local/bin/transmission-daemon -f --config-dir=/etc/transmission-daemon --log-info 
ExecStop=/bin/kill -s STOP $MAINPID
ExecReload=/bin/kill -s HUP $MAINPID

[Install]
WantedBy=multi-user.target
```

**Reply by MAD-BUG** [reply]
For myself i found changing service type to exec fixes it. According to systemd.service man "Behavior of notify is similar to exec; however, it is expected that the service sends a "READY=1" notification message via sd_notify or an equivalent call when it has finished starting up.", but for some reason it doesn't work

**Reply by nevack** [reply]
Could you try increasing default timeout?
There's a chance this line is not reached before default systemd timeout is triggered
https://github.com/transmission/transmission/blob/586cff9506487eabf72baa3c1588e623a8b1ac63/daemon/daemon.cc#L848

Also, daemon must be built with SystemD support for this to work.

**Reply by remd** [reply]
> Also, daemon must be built with SystemD support for this to work.

What does this mean exactly? I just built 4.0.4 on Ubuntu 22.04.1 LTS following the dependency instructions [here](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#ubuntu). I followed the build instructions [here](https://github.com/transmission/transmission#building-a-transmission-release-from-the-command-line). The dependency instructions don't make any mention of requiring libsystemd-dev.

My `/lib/systemd/system/transmission-daemon.service` looks like this:
```
[Unit]
Description=Transmission BitTorrent Daemon
Wants=network-online.target
After=network-online.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/local/bin/transmission-daemon -f --log-level=error
ExecReload=/bin/kill -s HUP $MAINPID
NoNewPrivileges=true
MemoryDenyWriteExecute=true
ProtectSystem=true
PrivateTmp=true

[Install]
WantedBy=multi-user.target
```

During startup my syslog looks like this:
```
Nov 30 00:21:46 knox systemd[1]: Starting Transmission BitTorrent Daemon...
Nov 30 00:23:16 knox systemd[1]: transmission-daemon.service: start operation timed out. Terminating.
Nov 30 00:23:21 knox transmission-daemon[3042629]: Closing transmission session... done.
Nov 30 00:23:21 knox systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
Nov 30 00:23:21 knox systemd[1]: Failed to start Transmission BitTorrent Daemon.
```

There's handful of unregistered torrents that might be slowing down startup time but I wasn't having this problem on 3.0.0.

**Reply by remd** [reply]
Fixed my issue by uninstalling transmission, installing libsystemd-dev, and rebuilding after purging the build/ folder.

I think libsystemd-dev should be a dependency mentioned in the package lists at https://github.com/transmission/transmission/blob/e10689beeabbf70b54cd0c2df8b4f0008fe05fe2/docs/Building-Transmission.md

After reinstalling with the package built using libsystemd-dev systemctl is notified quickly after startup that the startup was successful. I don't think increasing the timeout would work since the call to sd_notify would never work without that dependency available in the build.

```
Nov 30 01:02:30 knox systemd[1]: Starting Transmission BitTorrent Daemon...
Nov 30 01:02:32 knox systemd[1]: Started Transmission BitTorrent Daemon.
```

**Reply by killemov** [reply]
@ingria It's right there: `TimeoutStartSec=infinity`

**Comment by MaartenDeen** [comment]
Can confirm @remd . Had the same error with timeout, installed libsystemd-dev, did a rebuild and worked like a charm. 4.0.4 on Ubuntu 22.04.3

**Comment by 07freedom** [comment]
Confirm @remd. I reinstall with 4.0.4 with libsystemd-dev and change all config files with privilege 777. It worked, finally.

**Comment by disconnect5852** [comment]
so i have to find another torrent client (i didn't built anything, and wont be, installed from apt)? I've upgraded from ubuntu 22.04 LTS to 24.04 LTS, and i had such errors. Then did a chown for debian-transmission user and group on the config files.
The permission denied messages are gone, now not error, just fails to start the same way (after about a minute waiting).

`myuser@newnas:~$ journalctl -xeu transmission-daemon.service
░░ 
░░ The unit transmission-daemon.service completed and consumed the indicated resources.
Nov 18 15:07:15 newnas systemd[1]: Starting transmission-daemon.service - Transmission BitTorrent Daemon...
░░ Subject: A start job for unit transmission-daemon.service has begun execution
░░ Defined-By: systemd
░░ Support: http://www.ubuntu.com/support
░░ 
░░ A start job for unit transmission-daemon.service has begun execution.
░░ 
░░ The job identifier is 2265.
Nov 18 15:08:45 newnas systemd[1]: transmission-daemon.service: start operation timed out. Terminating.
Nov 18 15:09:01 newnas transmission-daemon[6691]: Closing transmission session... done.
Nov 18 15:09:01 newnas systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: http://www.ubuntu.com/support
░░ 
░░ The unit transmission-daemon.service has entered the 'failed' state with result 'timeout'.
Nov 18 15:09:01 newnas systemd[1]: Failed to start transmission-daemon.service - Transmission BitTorrent Daemon.
░░ Subject: A start job for unit transmission-daemon.service has failed
░░ Defined-By: systemd
░░ Support: http://www.ubuntu.com/support
░░ 
░░ A start job for unit transmission-daemon.service has finished with a failure.
░░ 
░░ The job identifier is 2265 and the job result is failed.
Nov 18 15:09:01 newnas systemd[1]: transmission-daemon.service: Consumed 16.155s CPU time.
░░ Subject: Resources consumed by unit runtime
░░ Defined-By: systemd
░░ Support: http://www.ubuntu.com/support
░░ 
░░ The unit transmission-daemon.service completed and consumed the indicated resources.
`

**Comment by remd** [comment]
I've been seeing this crop up again after upgrading to Ubuntu 24.04.1 LTS and using apt-installed Transmission 4.0.5. It seems to only happen on system restart. Manually starting the service does not encounter a timeout issue.

```
2025-05-08T09:54:06.675105-04:00 knox systemd[1]: Starting transmission-daemon.service - Transmission BitTorrent Daemon...
2025-05-08T09:55:36.691669-04:00 knox systemd[1]: transmission-daemon.service: start operation timed out. Terminating.
...
2025-05-08T09:55:47.602918-04:00 knox kernel: audit: type=1400 audit(1746712547.601:8803): apparmor="ALLOWED" operation="rename_src" class="file" profile="transmission->
2025-05-08T09:55:47.602952-04:00 knox kernel: audit: type=1400 audit(1746712547.601:8804): apparmor="ALLOWED" operation="rename_dest" class="file" profile="transmission>
2025-05-08T09:55:47.602958-04:00 knox kernel: audit: type=1400 audit(1746712547.601:8805): apparmor="ALLOWED" operation="open" class="file" profile="transmission-daemon>
2025-05-08T09:55:47.650864-04:00 knox kernel: audit: type=1400 audit(1746712547.649:8806): apparmor="ALLOWED" operation="rename_src" class="file" profile="transmission->
2025-05-08T09:56:01.967920-04:00 knox kernel: kauditd_printk_skb: 469 callbacks suppressed
2025-05-08T09:56:01.967952-04:00 knox kernel: audit: type=1400 audit(1746712561.966:9276): apparmor="ALLOWED" operation="open" class="file" profile="transmission-daemon>
2025-05-08T09:56:05.261461-04:00 knox transmission-daemon[1351]: Closing transmission session... done.
2025-05-08T09:56:05.278670-04:00 knox systemd[1]: transmission-daemon.service: Failed with result 'timeout'.
2025-05-08T09:56:05.278803-04:00 knox systemd[1]: Failed to start transmission-daemon.service - Transmission BitTorrent Daemon.
2025-05-08T09:56:05.278940-04:00 knox systemd[1]: transmission-daemon.service: Consumed 31.400s CPU time.
```

You can see that AppArmor is allowing operations even after the termination is called - so it seems to me that Transmission is simply working on starting up and the timeout needs to be set longer. Is this something that Transmission can do during apt installation?

I overrode TimeoutStartSec to 5 minutes to hopefully prevent this issue in the future. 

I'm not entirely sure why startup during system boot is so much slower than manual startup after the system finishes booting. There's no timeout issue when I manually start the service and I don't understand why there is a difference in startup time.
